### PR TITLE
Slick/JDBC: Support `PreparedStatement` use in Java DSL

### PR DESCRIPTION
**Pull request by ihostage**
Extend Java DSL for using `java.sql.PreparedStatement`

References #2247

### CODE REVIEW

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala [hunk 1]
```diff
@@ -221,4 +365,11 @@ object Slick {
   private def toDBIO[T](javaDml: JFunction[T, String]): T => DBIO[Int] = { t =>
     SQLActionBuilder(javaDml.asScala(t), SetParameter.SetUnit).asUpdate
   }
+
+  private def toDBIO[T](javaDml: JBiFunction[T, Connection, PreparedStatement]): T => DBIO[Int] = { t =>
+    SimpleJdbcAction { ctx =>
```
Comment (ihostage) [inline slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:370]: I need help with this. I do not very know a Slick and I'm not sure about transactionality every `SimpleJdbcAction`.
It works in tests and our application, but I'm not sure that it's absolutely correct.

**Comment by ennru** [comment]
Thank you for suggesting this, it's a great improvement for the Java DSL.

I wonder if there is a way to create the prepared statement only once and update it for every element.

Maybe @renatocaval can have a look?

**Comment by ihostage** [comment]
> I wonder if there is a way to create the prepared statement only once and update it for every element.

Yep, I thought about that too. But `PreparedStatement` bound to `Connection` and It's a really problem for us 😞 
https://stackoverflow.com/questions/4647476/reusing-of-a-preparedstatement-between-methods

Reuse `PreparedStatement` very easy, when we have a loop and one connection 😄 But it's not our case 🤷‍♂️ 
Besides, the user can use an external `Map<Connection, PreparedStatement>` for storing statements and reuse them if it really wants to do that.

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala [hunk 2]
```diff
@@ -88,6 +109,30 @@ object Slick {
...
+   *                    DDL statement is acceptable.
+   */
+  def flow[T](
+      session: SlickSession,
+      parallelism: Int,
+      toStatement: JBiFunction[T, Connection, PreparedStatement]
+  ): Flow[T, java.lang.Integer, NotUsed] =
+    ScalaSlick
+      .flow[T](parallelism, toDBIO(toStatement))(session)
+      .map(Int.box(_))
```
Comment (renatocaval) [inline slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:133]:
```suggestion
      .map(Int.box)
```

(same as above)

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala [hunk 3]
```diff
@@ -88,6 +109,30 @@ object Slick {
       .map(Int.box(_))
```
Comment (renatocaval) [inline slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:109]:
```suggestion
      .map(Int.box)
```

This is giving warning in IntelliJ, not essential though

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala (hunk 1 above)
Comment (renatocaval) [reply slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:370]: I don't get your question about transactionality. 

This DBIO will have a single statement and will run in its own transaction and not participating to any other transaction. If the user had access to it they could `flatMap` with another DBIO. In that case, they will need to add `.transactionally` to wrap both DBIO on the same tx. 

Anyway, that's not applicable here. 

However, what could be an addition is a `BiFunction[T, Connection, List[PreparedStatement]]`. 

The user could then return a few prepared statements, all bound to the same connection. 

In such a case, we would convert each of them to a DBIO and indeed execute them in the same transaction using `.transactionally`.

Can be added on another PR though.

Comment (ihostage) [reply slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:370]: > This DBIO will have a single statement and will run in its own transaction and not participating to any other transaction.

Exactly this is what I wanted to clarify. Thank you! 👍

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala (hunk 3 above)
Comment (ihostage) [reply slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala]: ✅ Done

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala (hunk 2 above)
Comment (ihostage) [reply slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala]: ✅ Done

File: docs/src/main/paradox/slick.md [hunk 4]
```diff
@@ -158,6 +158,8 @@ Scala
 Java
 : @@snip [snip](/slick/src/test/java/docs/javadsl/DocSnippetFlow.java) { #flow-example }
 
+Java (JDBC)
```
Comment (ennru) [inline docs/src/main/paradox/slick.md:161]: Please remove the old Java example. It is very bad to show this dangerous way of construction SQL statements.

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala [hunk 5]
```diff
@@ -64,6 +68,23 @@ object Slick {
...
     flow(session, 1, toStatement)
 
+  /**
+   * Java API: creates a Flow that takes a stream of elements of
+   *           type T, transforms each element to a SQL statement
+   *           using the specified function, and then executes
+   *           those statements against the specified Slick database.
+   *
+   * @param session The database session to use.
+   * @param toStatement A function that creeates the SQL statement to
```
Comment (ennru) [inline slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:78]:
```suggestion
   * @param toStatement A function that creates the SQL statement to
```
(same error on the method above BTW)

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala [hunk 6]
```diff
@@ -221,4 +365,11 @@ object Slick {
   private def toDBIO[T](javaDml: JFunction[T, String]): T => DBIO[Int] = { t =>
     SQLActionBuilder(javaDml.asScala(t), SetParameter.SetUnit).asUpdate
   }
+
+  private def toDBIO[T](javaDml: JBiFunction[T, Connection, PreparedStatement]): T => DBIO[Int] = { t =>
+    SimpleJdbcAction { ctx =>
+      val statement = javaDml(t, ctx.connection)
+      if (statement != null) statement.executeUpdate() else 0
```
Comment (ennru) [inline slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:372]: This becomes a hidden `filter` feature, but makes sense as a safe-guard.

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala [hunk 7]
```diff
@@ -64,6 +68,23 @@ object Slick {
...
+   *           those statements against the specified Slick database.
+   *
+   * @param session The database session to use.
+   * @param toStatement A function that creeates the SQL statement to
+   *                    execute from the current element. Any DML or
+   *                    DDL statement is acceptable.
+   */
+  def flow[T](
+      session: SlickSession,
+      toStatement: JBiFunction[T, Connection, PreparedStatement]
```
Comment (ennru) [inline slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:84]: Use `akka.japi.Function2` instead, so the `SQLException` can be thrown and fail the stream properly.

File: docs/src/main/paradox/slick.md (hunk 4 above)
Comment (ihostage) [reply docs/src/main/paradox/slick.md:161]: If we want not to show this approach in docs, maybe needly also to mark these API as deprecated? WDYT?

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala (hunk 5 above)
Comment (ihostage) [reply slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:78]: OMG. I fixed this in other places, but itself copy&past incorrect version 🤦

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala (hunk 6 above)
Comment (ihostage) [reply slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:372]: Yep. I thought about NPE-safety and I haven’t come up with anything better like this 🤷‍♂️

File: docs/src/main/paradox/slick.md (hunk 4 above)
Comment (ennru) [reply docs/src/main/paradox/slick.md:161]: It is still fine if the user knows what they are doing. Eg. when the data is properly typed or validated.
We can leave it as is, but not recommend it.

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala (hunk 6 above)
Comment (ennru) [reply slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala:372]: But it becomes a dangerous mix with the failure handling you suggested in the example...

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala (hunk 7 above)
Comment (ihostage) [reply slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala]: ✅ Done

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala (hunk 5 above)
Comment (ihostage) [reply slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala]: ✅ Done

File: docs/src/main/paradox/slick.md (hunk 4 above)
Comment (ihostage) [reply docs/src/main/paradox/slick.md:161]: ✅ Done

File: slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala (hunk 6 above)
Comment (ihostage) [reply slick/src/main/scala/akka/stream/alpakka/slick/javadsl/Slick.scala]: I removed this check, rewritten examples in docs, and added test with `recover` function.
Please, review.

**Comment by ennru** [comment]
Thank you @ihostage!

**Pull request merged**